### DISCUSSION TITLE
Enable prioritisiation: vcpkg, pkg-config, non-vcpkg find_package

### DISCUSSION THREAD

**Discussion by SamuelMarks** [microsoft/vcpkg#35659] [Q&A]
Sometimes I want a package from the OS (e.g., OpenSSL). Sometimes I want a package from vcpkg (e.g., brotli), and sometimes from pkg-config (e.g., libbsd).

I could imagine CMake `option`s like `PREFER_CONAN`, `PREFER_VCPKG`, `PREFER_OS`, `PREFER_PKG_CONFIG`.

Or maybe something that more explicitly specifies the priorities, with [your own](https://stackoverflow.com/a/39671055) [enum](https://gitlab.kitware.com/cmake/cmake/-/issues/23601), with a default priority list that can be overriden. Somethin like:
`conan; pkg-config; OS-find-package; vcpkg`

My use case is that I prioritise whatever is available on the target system, and only want to include my own shared/static variants if I know the target is guaranteed not to have this library.

EDIT - Here is one approach, including a custom `FindPackageName.cmake` for each `PackageName`:
```cmake
# - Find pcre (Perl Compatible Regular Expressions)
# Find the native PCRE2 headers and libraries.
#
# PCRE2_INCLUDE_DIRS    - where to find pcre.h, etc.
# PCRE2_LIBRARIES   - List of libraries when using pcre.
# PCRE2_FOUND   - True if pcre found.
include(FindPackageHandleStandardArgs)

set(SEARCH_PATH_FOR_INCLUDE "")
set(SEARCH_PATH_FOR_LIBRARY "")
if (DEFINED VCPKG_INSTALLED_DIR)
	foreach (p ${VCPKG_CMAKE_FIND_ROOT_PATH})
		list(APPEND SEARCH_PATH_FOR_INCLUDE "${p}/include")
		list(APPEND SEARCH_PATH_FOR_LIBRARY "${p}/lib")
	endforeach (p ${VCPKG_CMAKE_FIND_ROOT_PATH})
endif (DEFINED VCPKG_INSTALLED_DIR)
list(APPEND SEARCH_PATH_FOR_INCLUDE
		"/usr/include"
		"/usr/local/include"
		"/opt/local/include"
		"/sw/include"
		"${CMAKE_INCLUDE_PATH}"
		"${CMAKE_INSTALL_PREFIX}/include"
)
list(APPEND SEARCH_PATH_FOR_LIBRARY
		"/usr/lib"
		"/usr/lib64"
		"/usr/local/lib"
		"/usr/local/lib64"
		"/opt/local/lib"
		"/sw/lib"
		"${CMAKE_LIBRARY_PATH}"
		"${CMAKE_INSTALL_PREFIX}/lib"
)
if(PCRE2_LIBRARIES AND PCRE2_INCLUDE_DIRS)
	# in cache already
	set(PCRE2_FOUND TRUE)
else()
	find_path(PCRE2_INCLUDE_DIR
			NAMES
			pcre2.h
			PATHS
			${SEARCH_PATHS_FOR_INCLUDE})

	# Look for the library.
	find_library(PCRE2_LIBRARY
			NAMES
			libpcre2.a
			pcre2-8
			libpcre2-posix
			PATHS
			${SEARCH_PATHS_FOR_LIBRARY})

	if(PCRE2_INCLUDE_DIR AND PCRE2_LIBRARY)
		# learn pcre2 version
		file(STRINGS ${PCRE2_INCLUDE_DIR}/pcre2.h PCRE2_VERSION_MAJOR
				REGEX "#define[ ]+PCRE2_MAJOR[ ]+[0-9]+")
		string(REGEX MATCH " [0-9]+" PCRE2_VERSION_MAJOR ${PCRE2_VERSION_MAJOR})
		string(STRIP "${PCRE2_VERSION_MAJOR}" PCRE2_VERSION_MAJOR)

		file(STRINGS ${PCRE2_INCLUDE_DIR}/pcre2.h PCRE2_VERSION_MINOR
				REGEX "#define[ ]+PCRE2_MINOR[ ]+[0-9]+")
		string(REGEX MATCH " [0-9]+" PCRE2_VERSION_MINOR ${PCRE2_VERSION_MINOR})
		string(STRIP "${PCRE2_VERSION_MINOR}" PCRE2_VERSION_MINOR)

		set(PCRE2_VERSION ${PCRE2_VERSION_MAJOR}.${PCRE2_VERSION_MINOR})
	endif()

	set(PCRE2_INCLUDE_DIRS ${PCRE2_INCLUDE_DIR})
	set(PCRE2_LIBRARIES ${PCRE2_LIBRARY})
	mark_as_advanced(PCRE2_INCLUDE_DIRS PCRE2_LIBRARIES)

	# Handle the QUIETLY and REQUIRED arguments and set PCRE2_FOUND to TRUE if all listed variables are TRUE.
	find_package_handle_standard_args(PCRE2 FOUND_VAR PCRE2_FOUND
			REQUIRED_VARS PCRE2_LIBRARY PCRE2_INCLUDE_DIR
			VERSION_VAR PCRE2_VERSION)
endif()
```

Is this a good way? - Obviously some generalisation can be made with helper cmake functions so: each one of these files is like 4 lines; and you one could implement a simple parser to allow you to prioritise (which under the hood would rejig the search path reprioritising [vcpkg, conan, OS pkg-config, OS find-package])… but is there a better way?

**Comment by dg0yt** [comment]
There is only one reliable way: Use empty overlay ports where the implementation is not meant to come from vcpkg.
All source code must see the same (header and binary) version of the dependencies at build time and at runtime.
